File: softeer/택배마스터광우.java
import java.util.*;
import java.io.*;

class Main{
    static int n,m,k;
    static int rail[];
    static int matrix[];
    static boolean visited[];
    static int answer = Integer.MAX_VALUE;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        String[] inp = br.readLine().split(" ");
        n = Integer.parseInt(inp[0]);
        m = Integer.parseInt(inp[1]);
        k = Integer.parseInt(inp[2]);
        rail = new int[n];
        visited = new boolean[n];
        matrix = new int[n];

        inp = br.readLine().split(" ");
        
        for(int i = 0; i < n; i++){
            rail[i] = Integer.parseInt(inp[i]);
        }

        DFS(0);

        bw.append(answer + "\n");
        bw.flush();
        br.close();
        bw.close();
    }

    public static void DFS(int cnt){
        if(cnt == n){
            int idx = 0, sum = 0;

            for(int i = 0; i < k; i++){
                int temp = 0;
                while(temp + matrix[idx] <= m){
                    temp += matrix[idx];
                    idx = (idx+1)%n;
                }
                sum += temp;
            }
            if(answer > sum) answer = sum;
            return;
        }        

        for(int i = 0; i < n; i++){
            if(visited[i]) continue;

            visited[i] = true;
            matrix[cnt] = rail[i]; 
            DFS(cnt+1);
            visited[i] = false;
        }
    }
}

File: softeer/택배마스터광우.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <climits>

using namespace std;

int n,m,k;
int answer = INT_MAX;
vector<int> rail;

int main(){
    cin >> n >> m >> k;
    for(int i = 0; i < n; i++){
        int weight;
        cin >> weight;

        rail.push_back(weight);
    }

    sort(rail.begin(),rail.end());

    do{
        int cnt = 0, sum = 0, idx = 0;

        while(cnt < k){
            int temp = 0;
            while(temp + rail[idx] <= m){
                temp += rail[idx];
                idx = (idx+1)%n;
            }
            sum += temp;
            cnt++;
        }

        if(answer > sum) answer = sum;
    }while(next_permutation(rail.begin(),rail.end()));

    cout << answer << "\n";

    return 0;
}